megatron/checkpointing.py

```diff
@@ -211,6 +215,388 @@ def get_rng_state():
+    except BaseException as e:
+        print_rank_0('could not load the tensorizer checkpoint')
+        print_rank_0(e)
+        sys.exit()
+
+    return state_dict, release, checkpoint_name
+
+
+def flatten_dict_to_skeleton(d, parent_key=''):
+    flat, skel = {}, {}
+    if isinstance(d, dict):
+        for k, v in d.items():
+            new_key = f"{parent_key}.{k}" if parent_key else k
+            sub_flat, sub_skel = flatten_dict_to_skeleton(v, new_key)
+            flat.update(sub_flat)
+            skel[k] = sub_skel
+    elif isinstance(d, list):
+        skel_list = []
+        for idx, item in enumerate(d):
+            list_key = f"{parent_key}.{idx}"
+            sub_flat, sub_skel = flatten_dict_to_skeleton(item, list_key)
+            flat.update(sub_flat)
+            skel_list.append(sub_skel)
+        return flat, skel_list
+    else:
+        if isinstance(d, torch.Tensor):
+            flat[parent_key] = d
+            return flat, None
+        else:
+            return flat, d
+    return flat, skel
+
+
+
+def unflatten_to_skeleton(flat, skel):
+    for k, v in flat.items():
+        keys, d = k.split("."), skel
+        for key in keys[:-1]:
+            if isinstance(d, list):
+                d = d[int(key)]
+            elif isinstance(d, dict):
+                d = d.setdefault(key, {})
+        if isinstance(d, list):
+            d[int(keys[-1])] = v
+        else:
+            d[keys[-1]] = v
+    return skel
+
+
+def convert_parameters_to_tensors(d):
+    if isinstance(d, dict):
+        for k, v in d.items():
+            if isinstance(v, torch.nn.Parameter):
+                d[k] = v.data
+            else:
+                convert_parameters_to_tensors(v)
+    elif isinstance(d, list):
+        for idx, item in enumerate(d):
+            if isinstance(item, torch.nn.Parameter):
+                d[idx] = item.data
+            else:
+                convert_parameters_to_tensors(item)
+
+
+def serialize_model_state_dict(state_dict, stem):
+    for k, v in state_dict.items():
+        if isinstance(v, OrderedDict):
+            serializer = TensorSerializer(stream_io.open_stream(f"{stem}.{k}", 'wb+', buffer_size=0))
+            serializer.write_state_dict(v)
+            serializer.close()
+            state_dict[k] = None
+        elif isinstance(v, dict):
+            serialize_model_state_dict(v, stem)
+    return state_dict
+
+
+def dump_optimizer(opt, checkpoint_name):
+    flattened, skeleton = flatten_dict_to_skeleton(opt.state_dict())
+    serializer = TensorSerializer(f'{checkpoint_name}-opt.tensors')
+    serializer.write_state_dict(flattened)
+    serializer.close()
+    json.dump(skeleton, fp=open(f'{checkpoint_name}-opt.json', 'w'))
+
+
+def load_optimizer(checkpoint_name):
+    opt_state_dict = json.load(fp=open(f'{checkpoint_name}-opt.json', 'r'))
+    deserializer = TensorDeserializer(f'{checkpoint_name}-opt.tensors', device=torch.cuda.current_device(), plaid_mode=True)
```

Comment:
What are the column length standards for the megatron repo?

Reply:
It doesn't seem to be explicitly stated but the authors seem to style it consistently with [PEP 8 style guide](https://peps.python.org/pep-0008/#maximum-line-length) or max length of 79 characters. The original authors also use a [PEP 8 linter](https://github.com/NVIDIA/Megatron-LM/blob/fab0bd693ec5be55b32c4f12a1ea44766ec63448/tools/linter.py#L31).